File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/DynamicProgramming/120_Triangle.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


c.. Solution o..
    ___ minimumTotal  triangle
        """
        dp[i][j] is the j-th position's minimum path sum in i-th row.
        Then we can find the recursive formula:
            dp[i][j] = min(dp[i-1][j-1] if j-1>=0, dp[i-1][j]) + triangle[i][j]
        """

        __ n.. triangle or n.. triangle[0]:
            r_ 0
        rows = l..(triangle)
        dp = [0 ___ i __ xrange(rows)]

        ___ lines __ triangle:
            # Scan from tail to head to reduce space, otherwise we need a
            # rows*rows array.
            ___ j __ xrange(l..(lines) - 1, -1, -1
                __ (j - 1 >= 0 and dp[j - 1] < dp[j]) or j __ l..(lines) - 1:
                    dp[j] = dp[j - 1] + lines[j]
                ____
                    dp[j] += lines[j]

        r_ min(dp)


"""
[]
[[-10]]
[[2],[3,4],[6,5,7],[1,4,8,3]]
[[2],[3,3],[1,5,0],[4,6,9,3]]
"""

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/DynamicProgramming/120_Triangle.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-10-09 14:22:17
 */

c.. Solution {
public:
    /*
    dp[i][j] is the j-th position's minimum path sum in i-th row.
    Then we can find the recursive formula:
        dp[i][j] = min(dp[i-1][j-1] if j-1>=0, dp[i-1][j]) + triangle[i][j]
    */
    int minimumTotal(vector<vector<int>>& triangle) {
        int rows = triangle.size();
        __(rows__0 || triangle[0].size()__0){
            r_ 0;
        }

        std::vector<int> dp(rows, 0);
        ___(auto lines : triangle){
            ___(int i=lines.size()-1; i!=-1; i--){
                __(i-1>=0 && dp[i-1]<dp[i] || (i__lines.size()-1)){
                    dp[i] = dp[i-1] + lines[i];
                }
                else{
                    dp[i] += lines[i];
                }
            }
        }
        int min_path_sum = dp[0];
        ___(int i=1; i<rows; i++){
            min_path_sum = dp[i] < min_path_sum ? dp[i] : min_path_sum;
        }
        r_ min_path_sum;
    }
};

/*
[]
[[-10]]
[[2],[3,4],[6,5,7],[1,4,8,3]]
[[2],[3,3],[1,5,0],[4,6,9,3]]
*/
